include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -1,405 +1,969 @@
-    
-    float LplusV_rcpLen;
-    retval = createForReflection(tangentSpaceV.z,tangentSpaceL.z,VdotL,LplusV_rcpLen);
-    retval.TdotH = (tangentSpaceV.x+tangentSpaceL.x)*LplusV_rcpLen;
-    retval.BdotH = (tangentSpaceV.y+tangentSpaceL.y)*LplusV_rcpLen;
-    
-    return retval;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static AnisotropicMicrofacetCache createForReflection(
-    const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample)
-  {
-    return createForReflection(interaction.getTangentSpaceV(),_sample.getTangentSpaceL(),_sample.VdotL);
-  }
-  // transmissive cases need to be checked if the path is valid before usage
-  static bool compute(
-    out AnisotropicMicrofacetCache retval,
-    const bool transmitted, const float3 V, const float3 L,
-    const float3 T, const float3 B, const float3 N,
-    const float NdotL, const float VdotL,
-    const float orientedEta, const float rcpOrientedEta, out float3 H
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,transmitted,V,L,N,NdotL,VdotL,orientedEta,rcpOrientedEta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(T,H);
-      retval.BdotH = dot(B,H);
-    }
-    return valid;
-  }
-  template<class ObserverRayDirInfo, class IncomingRayDirInfo>
-  static bool compute(
-    out AnisotropicMicrofacetCache retval,
-    const surface_interactions::Anisotropic<ObserverRayDirInfo> interaction, 
-    const LightSample<IncomingRayDirInfo> _sample,
-    const float eta
-  )
-  {
-    float3 H;
-    const bool valid = IsotropicMicrofacetCache::compute(retval,interaction,_sample,eta,H);
-    if (valid)
-    {
-      retval.TdotH = dot(interaction.T,H);
-      retval.BdotH = dot(interaction.B,H);
-    }
-    return valid;
-  }
-
-  float TdotH;
-  float BdotH;
+    static this_t create(
+        const vector3_type tangentSpaceV,
+        const vector3_type tangentSpaceH,
+        const bool transmitted,
+        NBL_CONST_REF_ARG(fresnel::OrientedEtaRcps<monochrome_type>) rcpOrientedEta
+    )
+    {
+        this_t retval = createForReflection(tangentSpaceV,tangentSpaceH);
+        if (transmitted)
+        {
+            Refract<scalar_type> r = Refract<scalar_type>::create(tangentSpaceV, tangentSpaceH);
+            retval.iso_cache.LdotH = r.getNdotT(rcpOrientedEta.value2[0]);
+        }
+
+        return retval;
+    }
+    // always valid because its specialized for the reflective case
+    static this_t createForReflection(const vector3_type tangentSpaceV, const vector3_type tangentSpaceL, const scalar_type VdotL)
+    {
+        this_t retval;
+
+        scalar_type LplusV_rcpLen;
+        retval.iso_cache = isocache_type::createForReflection(tangentSpaceV.z, tangentSpaceL.z, VdotL, LplusV_rcpLen);
+        retval.TdotH = (tangentSpaceV.x + tangentSpaceL.x) * LplusV_rcpLen;
+        retval.BdotH = (tangentSpaceV.y + tangentSpaceL.y) * LplusV_rcpLen;
```

Comment:
care is needed, `LplusV_rcpLen` should be made negative if needed by the `isocache_type::createForReflection` to keep `NdotH` positive